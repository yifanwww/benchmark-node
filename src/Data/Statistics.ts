import { ConfidenceInterval, ConfidenceLevel, Mathematics } from '../Mathematics';
import { Arguments } from '../Parameterization';
import { ConsoleLogger } from '../Tools/ConsoleLogger';
import { Nanosecond } from '../types';
import { Optional } from '../types.internal';

/**
 * Class for stats including mean, margin or error, and standard deviation.
 */
export class Statistics {
    private declare _name: string;

    private declare _params: Optional<readonly unknown[]>;
    private declare _args?: Arguments;

    private declare _n: number;

    private declare _mean: Nanosecond;
    private declare _variance: number;
    private declare _standardDeviation: Nanosecond;
    private declare _standardError: Nanosecond;
    private declare _standardErrorPercent: number;

    private declare _ci: ConfidenceInterval;

    private declare _q0: Nanosecond;
    private declare _q1: Nanosecond;
    private declare _q2: Nanosecond;
    private declare _q3: Nanosecond;
    private declare _q4: Nanosecond;

    private declare _ops: number;

    get name() {
        return this._name;
    }

    get params() {
        return this._params;
    }

    get args() {
        return this._args;
    }

    /**
     * The measurements count.
     */
    get n() {
        return this._n;
    }

    /**
     * The arithmetic mean of measurements in nanoseconds.
     */
    get mean() {
        return this._mean;
    }

    /**
     * The variance of measurements.
     */
    get variance() {
        return this._variance;
    }

    /**
     * The standard deviation of measurements.
     */
    get standardDeviation() {
        return this._standardDeviation;
    }

    /**
     * The standard error of measurements in nanoseconds.
     */
    get standardError() {
        return this._standardError;
    }

    get confidenceInterval() {
        return this._ci;
    }

    get min() {
        return this._q0;
    }

    get q1() {
        return this._q1;
    }

    get median() {
        return this._q2;
    }

    get q3() {
        return this._q3;
    }

    get max() {
        return this._q4;
    }

    /**
     * The number of executions per second.
     */
    get ops(): number {
        return this._ops;
    }

    constructor(
        name: string,
        measurements: Nanosecond[],
        ops: number,
        params: Optional<readonly unknown[]>,
        args?: Arguments,
    ) {
        this._name = name;

        this._params = params;
        this._args = args;

        measurements.sort((a, b) => a - b);
        measurements = measurements.map((used) => used / ops);

        this._n = measurements.length;

        this._mean = Mathematics.mean(measurements);
        this._variance = Mathematics.variance(measurements, this._mean);
        this._standardDeviation = Math.sqrt(this._variance);
        this._standardError = this._standardDeviation / Math.sqrt(this._n);
        this._standardErrorPercent = (this._standardError / this._mean) * 100;

        this._ci = new ConfidenceInterval(this._n, this._mean, this._standardError, ConfidenceLevel.L95);

        this._q0 = measurements[0];
        this._q1 = measurements[Math.round(this._n * 0.25)];
        this._q2 = measurements[Math.round(this._n * 0.5)];
        this._q3 = measurements[Math.round(this._n * 0.75)];
        this._q4 = measurements[this._n - 1];

        this._ops = 1e9 / this._mean;
    }

    log(): void {
        const logger = ConsoleLogger.default;

        logger.writeLineStatistic(
            [
                `Mean = ${this._mean.toFixed(3)} ns`,
                `StdErr = ${this._standardError.toFixed(3)} ns (${this._standardErrorPercent.toFixed(2)}%)`,
                `N = ${this._n}`,
                `StdDev = ${this._standardDeviation.toFixed(3)} ns`,
            ].join(', '),
        );

        logger.writeLineStatistic(
            [
                `Min = ${this._q0.toFixed(3)} ns`,
                `Q1 = ${this._q1.toFixed(3)} ns`,
                `Median = ${this._q2.toFixed(3)}`,
                `Q3 = ${this._q3.toFixed(3)} ns`,
                `Max = ${this._q4.toFixed(3)} ns`,
            ].join(', '),
        );

        const [lower, upper] = [this._ci.lower, this._ci.upper];
        const margin = this._ci.margin.toFixed(3);
        const marginPercent = this._ci.marginPercent.toFixed(2);
        logger.writeLineStatistic(
            [
                `ConfidenceInterval = [${lower.toFixed(3)} ns; ${upper.toFixed(3)} ns] (CI ${this._ci.level * 100}%)`,
                `Margin = ${margin} ns (${marginPercent}% of Mean)`,
            ].join(', '),
        );

        logger.writeLine();
    }
}
