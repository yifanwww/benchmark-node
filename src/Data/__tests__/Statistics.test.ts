import { Statistics } from '../Statistics';

describe(`Test class \`${Statistics.name}\``, () => {
    it('calculates', () => {
        const stats = new Statistics(
            'name',
            [
                37051080, 33390980, 33920080, 33423180, 33684980, 34514380, 33498580, 34333480, 33923280, 34479480,
                33398680, 33242780, 33088980, 33790080, 33530180,
            ],
            141386309,
        );

        expect(JSON.stringify(stats, undefined, 4)).toMatchSnapshot();
    });
});
