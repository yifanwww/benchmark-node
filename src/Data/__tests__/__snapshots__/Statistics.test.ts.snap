// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test class \`Statistics\` calculates 1`] = `
Statistics {
  "_args": undefined,
  "_ci": ConfidenceInterval {
    "_degreeOfFreedom": 14,
    "_level": 0.95,
    "_lower": 0.23636101328984727,
    "_margin": 0.003770771087285532,
    "_marginPercent": 1.5702923696945692,
    "_upper": 0.24390255546441836,
  },
  "_mean": 0.2401317843771328,
  "_n": 15,
  "_name": "name",
  "_ops": 4164379999.0653286,
  "_q0": 0.2340324196453845,
  "_q1": 0.2363961562926153,
  "_q2": 0.23899117417373134,
  "_q3": 0.24283454489217907,
  "_q4": 0.26205564217678246,
  "_standardDeviation": 0.006809131045845519,
  "_standardError": 0.001758110076180239,
  "_standardErrorPercent": 0.7321438437400208,
  "_variance": 0.00004636426559949729,
}
`;

exports[`Test class \`Statistics\` calculates 2`] = `"name"`;

exports[`Test class \`Statistics\` calculates 3`] = `undefined`;

exports[`Test class \`Statistics\` calculates 4`] = `15`;

exports[`Test class \`Statistics\` calculates 5`] = `0.2401317843771328`;

exports[`Test class \`Statistics\` calculates 6`] = `0.00004636426559949729`;

exports[`Test class \`Statistics\` calculates 7`] = `0.006809131045845519`;

exports[`Test class \`Statistics\` calculates 8`] = `0.001758110076180239`;

exports[`Test class \`Statistics\` calculates 9`] = `
ConfidenceInterval {
  "_degreeOfFreedom": 14,
  "_level": 0.95,
  "_lower": 0.23636101328984727,
  "_margin": 0.003770771087285532,
  "_marginPercent": 1.5702923696945692,
  "_upper": 0.24390255546441836,
}
`;

exports[`Test class \`Statistics\` calculates 10`] = `0.2340324196453845`;

exports[`Test class \`Statistics\` calculates 11`] = `0.2363961562926153`;

exports[`Test class \`Statistics\` calculates 12`] = `0.23899117417373134`;

exports[`Test class \`Statistics\` calculates 13`] = `0.24283454489217907`;

exports[`Test class \`Statistics\` calculates 14`] = `0.26205564217678246`;

exports[`Test class \`Statistics\` calculates 15`] = `4164379999.0653286`;
